### PR TITLE
[rl] Enable multinode multi-generator RL and end2end agent run

### PR DESCRIPTION
**Pull request by yichuan-w**
## Enable multi-host, multi-generator RL (Search-R1, Qwen3-32B)

Builds on the multi-generator core (#3696, now in `main`) to make multinode RL actually
train a large model: a **multi-host FSDP trainer + N rollout generators on separate host
meshes**, plus the Qwen3-32B Search-R1 recipes that exercise it.

### What's in this PR (2 commits)

**1. Weight-sync transport knob + mixed-precision dtype fix** (`actors/{trainer,generator}.py`)
- `weight_sync_direct_rdma: bool | None` on both `PolicyTrainer.Config` and
  `VLLMGenerator.Config`. `None` = `is_rdma_available()`; `False` = CPU-staged via
  TorchStore StorageVolumes (lower trainer GPU memory); `True` = direct GPU-to-GPU RDMA.
  Trainer and generator must match. Direct RDMA fans out fine to multiple generators with
  a **single-host** trainer (validated: FSDP=8 + bf16 + 3 generators, ~4s weight sync vs
  ~133s CPU-staged, reward converges 0.21 -> ~0.42). The limitation is a **multi-host
  trainer**, where direct RDMA SIGSEGVs in Monarch's RDMA layer (`monarch_rdma`
  IbvMemoryRegion) -- multi-host trainers use `False` (the 2-host `_fsdp16` recipe stays
  CPU-staged).
- torchstore applies `put_state_dict(transfer_dtype=...)` **only** on the direct-RDMA path;
  the CPU-staged path ignores it. With mixed-precision FSDP the trainer keeps an **fp32
  master**, so its `state_dict()` is fp32 while the generator is bf16 -> the uncast weights
  trip the generator's pull-side dtype assert (`float32 != bfloat16`) before step 1. Fix:
  cast the published state_dict to the generator dtype on the CPU-staged path (direct-RDMA
  keeps using `transfer_dtype`).

**2. Qwen3-32B Search-R1 multinode recipes** (`examples/search_r1/config_registry.py`)
- `rl_grpo_qwen3_32b_search_r1`: 1-host FSDP=8 trainer + 3 generators (5 hosts).
- `rl_grpo_qwen3_32b_search_r1_fsdp16`: 2-host FSDP=16 trainer + 3 generators (6 hosts) --
  sharding the fp32 master + AdamW states 16-way halves per-rank memory so the faster
  SelectiveAC fits (vs FullAC at FSDP=8).
- Both: fp32 master + bf16 mixed-precision FSDP, decode-only cudagraphs (`FULL_DECODE_ONLY`,
  #3668), CPU-staged weight sync (`weight_sync_direct_rdma=False`), `num_generators=3`.
- Reconcile with `main`: document the `q_BLNH/k_BLNH/v_BLNH` arg convention in
  `VLLMAttentionWrapper`; drop the `disable_loss_parallel` kwargs left in `alphabet_sort`
  after #3694 removed that `ParallelismConfig` field.

### Validation (MAST, Qwen3-32B Search-R1)

2-host FSDP=16 trainer (fp32 master + bf16) + 3 generators (TP=8) + a dense retriever,
end-to-end. Trains cleanly past the historical step-1/step-2 OOM:

| step | rollout_reward/_mean | resp_len max | grad_norm | tokens/s |
|---|---|---|---|---|
| 1 | 0.19 | 91  | 1.28 | 1128 |
| 2 | 0.30 | 148 | 1.97 | 1049 |
| 3 | 0.29 | 137 | 2.66 | 1177 |

Reward rises from the Qwen3-32B Search-R1 baseline (~0.19) and the model genuinely searches
then answers (response length growing). (Driven by an internal MAST conda launcher that is
not part of this PR.)

### CODE REVIEW

**Comment by yichuan-w** [comment]
### Results: Qwen3-32B Search-R1, 1×TP8 trainer + 3×TP8 generators (5 hosts)

Running this topology end-to-end on MAST (1 controller + 1 trainer (bf16, TP=8) + 3 generators (TP=8), `--num_generators 3`). Trains cleanly, no OOM, ~80-140s/step; the per-step weight sync fans out to all 3 generators fine.

**EM (`validation_reward/_mean`, held-out NQ, every 5 steps)** — wandb run `noif1ir9`:

| step | 0 | 5 | 10 | 15 | 20 | 25 | 30 |
|---|---|---|---|---|---|---|---|
| EM | 0.254 | 0.310 | 0.388 | 0.332 | 0.368 | 0.400 | 0.416 |

Baseline 0.254 -> 0.416 by step 30 (+0.16, ~+64%); rising with the usual GRPO step-to-step noise. Still early (step ~31 of 500). For reference the 8B on the same recipe rose ~0.20 -> ~0.52 over 500 steps, and the 32B starts higher and is tracking ahead.

wandb: https://meta.wandb.io/yichuan/torchtitan/runs/noif1ir9 (curve continues to update).

**Comment by tianyu-l** [comment]
trainer should use FSDP, mixed precision, master weights fp32.

**Comment by yichuan-w** [comment]
### Engineering notes from bringing the 32B multinode run up end-to-end

A few non-obvious findings while getting the 2-host FSDP trainer + 3 generators to train,
in case they help the next multinode-RL bringup.

**Weight sync (what this PR fixes).** Four failure modes were on the table: (1) titan
misusing torchstore, (2) a conda/ABI mismatch, (3) a torchstore bug, (4) a Monarch RDMA
bug. Mapping the path settled it on the titan side: the trainer is correct -- it publishes
the **sharded FSDP DTensors** and torchstore reshards them (no `full_tensor()` gather). The
one real titan-side gap is the mixed-precision dtype: CPU-staged `put_state_dict` silently
ignores `transfer_dtype`, so an fp32 master reaches the bf16 generator and trips the
pull-side assert before step 1 -- hence the explicit cast in this PR. Multi-generator must
use CPU-staging (`weight_sync_direct_rdma=False`): direct RDMA is single-reader and its
transient spike OOMs large generators.

**FSDP fp32-master 32B trainer.** With `data_parallel_shard_degree=8` + SelectiveAC the run
hit an NCCL watchdog timeout that *looked* like an all-gather deadlock but was actually a
**CUDA OOM in `compute_logprobs` on rank 0**: it upcasts the full `[B*S, vocab=151936]`
logits to fp32 (~2.5 GB) and `cross_entropy` allocates another ~2.5 GB; with TP=1 the vocab
is unsharded so every rank materializes it, and rank 0 (extra logging residency) loses the
race. Two fixes, both in the recipes: FullAC at FSDP=8, or FSDP=16 (2 hosts) which halves
per-rank master+optimizer memory and lets the faster SelectiveAC fit. (Flight Recorder +
the per-rank structured logger sorted by `seq_id` localized it -- rank 0 logged
`compute_logprobs_error: OutOfMemory` then never issued its loss/backward collectives, so
the peers blocked on the next collective.)

**Search-R1 retriever (infra, not in this PR).** Load the e5 encoder with
`attn_implementation="eager"`: the default torch SDPA hits an intermittent cuDNN fused-MHA
graph failure (`mha_graph.execute(...).is_good() == false`) on the 80 GB GPUs that 500s
`/retrieve`. (faiss / corpus alignment were fine -- verified against the full 64 GB index.)

Validation curve (2-host FSDP=16 + 3 generators) is in the PR description: reward
0.19 -> 0.30 over the first steps, healthy grad norms, model searches then answers.

**Comment by yichuan-w** [comment]
### Multinode RL over InfiniBand: a weight-sync transport gotcha (+ a sizable step-time win)

Follow-up notes from running these recipes multi-host (a multi-host FSDP trainer + N rollout
generators on separate host meshes), in case they help others bringing up multinode RL. This is
infra around the recipes, not a change to this PR.

**InfiniBand is worth it for the cross-host legs.** Moving the cross-host traffic -- the FSDP
all-gather / reduce-scatter for the sharded trainer, plus the per-step weight sync -- from TCP to
IB/RDMA cut step time from ~231s to ~148s (roughly a third faster) on a 2-host FSDP=16 trainer + 3
generators. Rollout generation is transport-independent, so the win is entirely in the cross-host
collectives + the weight transfer; the reward curve is unchanged.

**Gotcha: the weight-sync RDMA transport vs the generator's GPU weight buffers.** Once IB is
actually up, torchstore auto-selects its torchcomms (CTRAN) RDMA transport for the weight sync. On
the pull side the generator hands its own (vLLM) GPU weight tensors to the transport, which eagerly
RDMA-registers that GPU memory. On a vLLM generator that registration can fail with
`commInternalError` -- the buffer is cuMem/VMM-backed and/or the GPU and NIC are not PCIe-close --
and it fails *before step 1*. It is independent of trainer host count (a 1-host FSDP=8 trainer and a
2-host FSDP=16 trainer fail identically). The MonarchRDMA / SharedMemory transports register the
same tensors fine, via a different registration path.

**Workaround that keeps IB's speed.** Route the *weight-sync transport* off torchcomms (so it uses
SharedMemory same-host / MonarchRDMA cross-host) while leaving NCCL itself on IB for the FSDP
collectives. FSDP keeps full RDMA speed, and the weight sync uses a path that does not hit the
GPU-buffer registration. The proper fix belongs upstream in torchstore/torchcomms -- e.g. staging
the registration buffer on host memory (the affinity-free path), or handling cuMem-backed buffers in
the GPU registration.

(The multinode launcher driving this is internal and not part of this PR.)

**Comment by yichuan-w** [comment]
### Full 500-step convergence run -- `rl_grpo_qwen3_32b_search_r1_fsdp16`

Ran the 2-host FSDP=16 trainer + 3 generators recipe end-to-end for the full 500 GRPO steps; the job completed cleanly (no OOM / NCCL hang) across the multinode trainer + multi-generator fanout.

**`rollout_reward/_mean`** (training reward, sampled every 50 steps):

| step | 1 | 50 | 100 | 150 | 200 | 250 | 300 | 350 | 400 | 450 | 500 |
|------|------|------|------|------|------|------|------|------|------|------|------|
| reward | ~0.21 | 0.38 | 0.56 | 0.60 | 0.48 | 0.57 | 0.51 | 0.57 | 0.57 | 0.51 | 0.48 |

**`validation_reward/_mean`** (held-out exact-match accuracy, sampled every 50 steps):

| step | 5 | 50 | 100 | 150 | 200 | 250 | 300 | 350 | 400 | 450 | 500 |
|------|------|------|------|------|------|------|------|------|------|------|------|
| acc | 0.41 | 0.41 | 0.46 | 0.48 | 0.50 | 0.43 | 0.50 | 0.52 | 0.50 | 0.52 | 0.48 |

Reward rises over the first ~100-150 steps then plateaus (noisy) around ~0.55 rollout / ~0.50 validation exact-match, up from ~0.2 / ~0.41 at the start. Confirms the multinode multi-generator recipe (CPU-staged weight sync, `weight_sync_direct_rdma=False`) trains stably to completion on Qwen3-32B.

**Comment by yichuan-w** [comment]
I think this PR is ready, tldr: add better RDMA use flag, add CPU- staging precision cast; add more multi node search R1 test cases.

**Comment by yichuan-w** [comment]
### Multinode Search-R1 (Qwen3-32B) -- config + perf comparison

Three configs run end-to-end (1 controller + trainer(s) + 3 generators, TP=8 each):

| config | weight sync | per-step | result |
|---|---|---|---|
| 1-host trainer (FSDP=8) + 3 gen, bf16 + **direct RDMA** | **~4s** | **~38s/step** | works, fastest; reward converges |
| 2-host trainer (FSDP=16) + 3 gen, **CPU-staged** (`direct_rdma=False`) | ~133s | ~179s/step | works (slower); ran the full 500 steps to convergence |
| 2-host trainer (FSDP=16) + 3 gen, **direct RDMA** | -- | -- | does not work: SIGSEGV in Monarch's RDMA layer at training start |

Per-step breakdown:

```
FSDP=8 + RDMA  (~38s/step)        FSDP=16 + CPU-staged  (~179s/step)
  weight sync   ~4s                 weight sync   ~133s   <- bottleneck
  rollout      ~20s                 rollout       ~37.5s
  train        ~15s                 train          ~7s
```

Takeaways:
- Direct GPU-to-GPU RDMA cuts the trainer->generator weight sync ~30x (133s -> 4s), so per-step drops ~4.7x and the bottleneck moves back to rollout generation.
- Multi-generator fanout over RDMA is fine -- 3 generators concurrently reading the trainer's RDMA buffers works (the earlier "single-reader / >1 generator OOMs" note was wrong; fixed in this PR).
- Direct RDMA only works with a single-host trainer. A 2-host (FSDP=16) trainer + direct RDMA SIGSEGVs inside Monarch's RDMA layer, so multi-host trainers stay on CPU-staged for now (tracked as a TODO on the `rl_grpo_qwen3_32b_search_r1_fsdp16_rdma` recipe).
